python/ray/data/dataset.py

```diff
@@ -845,14 +845,128 @@ def groupby(self, key: "GroupKeyT") -> "GroupedDataset[T]":
+
+            on a dataset with columns ``col1`` and ``col2`` would yield
+            aggregation columns::
+
+                col1_agg1, col2_agg1, col1_agg2, col2_agg2
+
+            When ``aggs`` is given as a dict, the aggregation columns are
+            ordered in accordance with the depth-first dict ordering
+            (cols, aggs)::
+
+                {col1: [agg1, agg2], col2: [agg2, agg1]}
+
+            would yield aggregation columns::
+
+                col1_agg1, col1_agg2, col2_agg2, col2_agg1
+
+            If the dataset is empty, return ``None``.
+        """
+        aggs_ = self._build_multi_agg(aggs)
+        return self.apply_aggregations(*aggs_)
+
+    def _build_multi_agg(
+            self,
+            aggs: Union[str, AggregateFn, List[Union[str, AggregateFn]], Dict[
+                str, Union[str, AggregateFn]]],
+            skip_cols: Optional[List[str]] = None) -> List[Aggregation]:
+        """Build set of aggregations for applying multiple aggregations to
+        one or more columns.
+        """
+        aggs_ = []
+        if isinstance(aggs, (str, AggregateFn)):
+            aggs = [aggs]
+        err_msg = (
+            "`aggs` must be an aggregation name, AggregateFn, list of "
+            "aggregation names, list of AggregateFn, or a column name --> "
+            f"list of aggregations dict; instead, we got {aggs}.")
+        if isinstance(aggs, list):
+            if len(aggs) == 0:
+                raise ValueError("List of aggregations must be nonempty.")
+            if not all(isinstance(agg, (str, AggregateFn)) for agg in aggs):
+                raise TypeError(err_msg)
+            # Add dummy null column selection; this will cause all columns
+            # to be selected.
+            aggs = ((None, aggs), )
+        elif isinstance(aggs, dict):
+            if len(aggs) == 0:
+                raise ValueError("Dict of aggregations must be nonempty.")
+            if not all(
+                    isinstance(agg, (str, AggregateFn))
+                    for aggs_ in aggs.values() for agg in aggs_):
+                raise TypeError(err_msg)
+            aggs = aggs.items()
+        else:
+            raise TypeError(err_msg)
+
+        for on, on_aggs in aggs:
+            on = self._check_and_normalize_agg_on(on, skip_cols=skip_cols)
+            if not isinstance(on, list):
+                on = [on]
+            for agg in on_aggs:
+                if isinstance(agg, str):
+                    try:
+                        agg = AGGS_TABLE[agg]()
+                    except KeyError as e:
+                        raise ValueError(
+                            f"{e!s} is not a supported aggregation: "
+                            f"{list(AGGS_TABLE.keys())}") from None
+                for on_ in on:
+                    agg_ = Aggregation(agg, on_)
+                    aggs_.append(agg_)
+        return aggs_
+
+    def apply_aggregations(self, *aggs: Aggregation) -> U:
```

Comment:
It's confusing to have both agg() and apply_aggregations(), can we combine them?